src/inmanta/db/versions/v202503030.py

```diff
@@ -0,0 +1,238 @@
+limitations under the License.
+
+Contact: code@inmanta.com
+"""
+
+import hashlib
+import os
+from collections import defaultdict
+from pathlib import Path
+
+from asyncpg import Connection
+
+from inmanta import const, loader
+
+
+async def update(connection: Connection) -> None:
+    """
+    * Create the module table, that keeps track of the python package requirements
+      per inmanta module (name, version).
+    * Create the files_in_module table, that keeps track of which files belong to which inmanta
+      module.
+    * Create the modules_for_agent table, that keeps track of which inmanta modules are required
+      by which agent.
+    """
+
+    schema = """
+        CREATE TABLE public.module (
+            name varchar NOT NULL,
+            version varchar NOT NULL,
+            environment uuid NOT NULL REFERENCES environment(id) ON DELETE CASCADE,
+            requirements varchar[] DEFAULT ARRAY[]::varchar[] NOT NULL,
+            PRIMARY KEY(environment, name, version)
+        );
+
+        CREATE TABLE public.files_in_module (
+            module_name varchar NOT NULL,
+            module_version varchar NOT NULL,
+            environment uuid NOT NULL,
+            file_content_hash varchar NOT NULL REFERENCES file(content_hash) ON DELETE CASCADE,
+            file_path varchar NOT NULL,
+            PRIMARY KEY(environment, module_name, module_version, file_path),
+            FOREIGN KEY (environment, module_name, module_version)
+                REFERENCES public.module(environment, name, version) ON DELETE CASCADE
+        );
+
+        CREATE INDEX files_in_module_file_content_hash_index
+        ON public.files_in_module (file_content_hash);
+
+        CREATE TABLE public.modules_for_agent (
+            cm_version integer NOT NULL,
+            agent_name varchar NOT NULL,
+            module_name varchar NOT NULL,
+            module_version varchar NOT NULL,
+            environment uuid NOT NULL,
+            PRIMARY KEY(environment, cm_version, agent_name, module_name, module_version),
+            FOREIGN KEY (environment, cm_version)
+                REFERENCES public.configurationmodel(environment, version) ON DELETE CASCADE,
+            FOREIGN KEY (environment, agent_name)
+                REFERENCES public.agent(environment, name) ON DELETE CASCADE,
+            FOREIGN KEY (environment, module_name, module_version)
+                REFERENCES public.module(environment, name, version) ON DELETE CASCADE
+        );
+
+        CREATE INDEX modules_for_agent_environment_agent_name_index
+        ON public.modules_for_agent (environment, agent_name);
+        CREATE INDEX modules_for_agent_environment_module_name_module_version_index
+        ON public.modules_for_agent (environment, module_name, module_version);
+
+    """
+
+    await connection.execute(schema)
+
+    code_data: dict[str, dict[int, dict[str, set[tuple[str, str, str, set[str]]]]]] = defaultdict(
```

Comment:
Typing is very unreadable. Maybe adding dataclasses would make it easier to read ?

Reply:
I would add least document what it is. Named tuples or dataclasses can help, but I would not invest too much effort

Reply:
The overall approach can also use some comments, as this is very hard to follow